### PR TITLE
[CTOR-438] [Plugin] [CommunityPR] Infor ion grid restapi

### PR DESCRIPTION
**Pull request by lucie-dubrunfaut**
## Description

Adds a new plugin to monitor Infor ION Grid components through its Rest API.
Development from C.Gagnaire [4944](https://github.com/centreon/centreon-plugins/pull/4944)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

### CODE REVIEW

**Review by sc979 — APPROVED** [review]

**Comment by cgagnaire** [comment]
Hi @sc979,
From where do you get this code style ?

**Comment by sc979** [comment]
Hi @cgagnaire ,
Not sure to understand 🤔 
Are you talking about the "Update coding style" commit (915eefb) ?

**Comment by cgagnaire** [comment]
Yes I am.

**Comment by sc979** [comment]
This is not coming from the pipeline team.
I will let @lucie-dubrunfaut reply to you

**Comment by cgagnaire** [comment]
Sorry, I've mistaken your approvement with the commit, its @lucie-dubrunfaut commit.
Anyway, someone needs to explain to me how this can be seriously considered:
image

**Comment by lucie-dubrunfaut** [comment]
Hi @cgagnaire , 
We are currently testing a common convergence of coding style in the connectors team. This is still progress and we are open to your feedback regarding this. I was just doing some internal feedback on this one.

**Comment by cgagnaire** [comment]
Hi @lucie-dubrunfaut,
No problem, I can say that I don't like it for now :)

**Comment by lucie-dubrunfaut** [comment]
I don't really like what's out yet either, but we had to start somewhere, and for now I had to redo a lot of things by hand. :(

**Comment by cgagnaire** [comment]
> I don't really like what's out yet either, but we had to start somewhere, and for now I had to redo a lot of things by hand. :(

What are you trying to achieve here ? What's the point of a coding style in this project ?

File: src/apps/infor/ion/grid/api/mode/application.pm [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    return "Total applications ";
+}
+
+sub set_counters {
+    my ($self, %options) = @_;
+
+    $self->{maps_counters_type} = [
+        { name             => 'global',
+          type             => 0,
+          cb_prefix_output => 'prefix_global_output' },
```
[the remark this answers is not in the document]
Comment (lucie-dubrunfaut) [reply src/apps/infor/ion/grid/api/mode/application.pm]: Not really it's a kind of mix between it and manual modifications

**Comment by lucie-dubrunfaut** [comment]
> > > > I don't really like what's out yet either, but we had to start somewhere, and for now I had to redo a lot of things by hand. :(
> > > 
> > > 
> > > What are you trying to achieve here ? What's the point of a coding style in this project ?
> > 
> > 
> > We try to harmonize the code that is maintained and added here. We used Pycharm's perl coding style parameters but at this moment as we have configured them, they seem not perfectly match. If you have suggestions for tools that would allow us to standardize our coding style, we are open to suggestions.
> 
> Well I don't know about Pycharm and I don't think we need anyone or anything to tell us how we should style our code :) The code I've pushed here https://github.com/centreon/centreon-plugins/pull/4965/files is I think a good example of how we should do. There is no need to be as strict as you did here, this is no PSR.

Okay I thought you were using an automatic code reformatting method to harmonize the coding style of this project. Thank you in any case for your feedback, we can discuss it internally.

On another topic, I added a "total" counter in application mode but I have not access to the API so I can't check myself if it is relevant. If the mode is only used with service discovery and therefore a filter-name, maybe not. But if it is possible to have several applications returned, this counter will potentially be useful for some users. What do you think?

**Review by Evan-Adam — DISMISSED** [review]

**Review by omercier — DISMISSED** [review]

**Review by Evan-Adam — DISMISSED** [review]

File: src/apps/infor/ion/grid/custom/restapi.pm [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+API username.
+
+=item B<--api-password>
+
+API password.
+
+=item B<--timeout>
+
+Threshold for HTTP timeout (default: '20').
```
Comment (omercier) [inline src/apps/infor/ion/grid/custom/restapi.pm:213]:
```suggestion
Define the name or the address of the host.

=item B<--port>

Define the port to connect to (default: '443').

=item B<--proto>

Select the right protocol between 'http' and 'https' (default: 'https').

=item B<--api-username>

Username to access the API.

=item B<--api-password>

Password to access the API.

=item B<--timeout>

Maximum time to wait (in seconds) for a response to HTTP requests before timeout (default: '20').
```

File: src/apps/infor/ion/grid/mode/application.pm [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+%{name} or %{description} and Regexp can be used.
+Typical syntax: --critical-status='%{started} ne "true"'
+
+=item B<--warning-total>
+
+Define the conditions to match for the total to be WARNING (default: '').
+
+=item B<--critical-total>
+
+Define the conditions to match for the total to be CRITICAL (default: '').
```
Comment (omercier) [inline src/apps/infor/ion/grid/mode/application.pm:177]:
```suggestion
=head1 MODE

Monitor the status of the application.

=over 8

=item B<--filter-name>

Define which applications should be monitored based on their names. This option will be treated as a regular expression.
Example: --filter-name='^application1$'

=item B<--warning-status>

Define the conditions to match for the status to be WARNING (default: '').
The condition can be written using special variables like %{state}, %{online}, %{started},
%{name} or %{description}. Regular expressions are supported.
Typical syntax: --warning-status='%{state} ne "OK"'

=item B<--critical-status>

Define the conditions to match for the status to be CRITICAL (default: '%{online} =~ /true/ && %{state} !~ /^(OK)/i').
The condition can be written using special variables like %{state}, %{online}, %{started},
%{name} or %{description}. Regular expressions are supported.
Typical syntax: --critical-status='%{started} ne "true"'

=item B<--warning-total>

Define the warning threshold for the total number of applications.

=item B<--critical-total>

Define the critical threshold for the total number of applications.
```

File: src/apps/infor/ion/grid/mode/node.pm [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+=item B<--critical-status>
+
+Set critical threshold for status.
+Can use special variables like: %{state}, %{name}, %{host_name},
+%{application_name}, %{type}.
+
+=item B<--warning-*> B<--critical-*>
+
+Thresholds.
+Can be: 'log-error', 'log-warning', 'uptime' (s), 'cpu-usage', 'heap-usage' (%).
```
Comment (omercier) [inline src/apps/infor/ion/grid/mode/node.pm:240]:
```suggestion
Monitor the status and the statistics of the nodes.

=over 8

=item B<--filter-type>

Define which nodes should be monitored based on the their type. This option will be treated as a regular expression.

=item B<--filter-name>

Define which nodes should be monitored based on the their name. This option will be treated as a regular expression.

=item B<--filter-application-name>

Define which applications should be monitored based on the their name. This option will be treated as a regular expression.

=item B<--filter-host-name>

Define which hosts should be monitored based on the their name. This option will be treated as a regular expression.

=item B<--warning-status>

Define the conditions to match to return a warning status (default: "%{state} !~ /online/").
The condition can be written using the following macros: %{state}, %{name}, %{host_name},
%{application_name}, %{type}.

=item B<--critical-status>

Define the conditions to match to return a critical status.
The condition can be written using the following macros: %{state}, %{name}, %{host_name},
%{application_name}, %{type}.

=item B<--warning-*> B<--critical-*>

Thresholds for 'log-error', 'log-warning', 'uptime' (s), 'cpu-usage', 'heap-usage' (%).
```

**Review by omercier — APPROVED** [review]

**Review by Evan-Adam — APPROVED** [review]

**Comment by cgagnaire** [comment]
This coding style is not good :)

**Comment by lucie-dubrunfaut** [comment]
> This coding style is not good :)

Hi @cgagnaire :)
Can you tell me precisely what you think is wrong so that I can take note of it and we can discuss it again on our side?

**Pull request merged**